File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const answer = [];
    const genresObj = {};
    const playsObj = {};
    
    for(let i = 0; i < genres.length; i++){
        if(genresObj[genres[i]] !== undefined) {
            genresObj[genres[i]] += plays[i]
            playsObj[genres[i]].push([plays[i], i])
        } else {
            genresObj[genres[i]] = plays[i]
            playsObj[genres[i]] = [[plays[i], i]]
        }
    }
    const sortedGenres = Object.keys(genresObj).sort((a, b) => genresObj[b] - genresObj[a]);
    for(const genre of sortedGenres){
        const sortedSongs = playsObj[genre].sort((a, b) => {
            return a[0] === b[0] ? a[1] - b[1] : b[0] - a[0];
        })
        answer.push(...sortedSongs.slice(0, 2).map(song => song[1]))
    }
    return answer
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    genre_play = {}
    genre_song = {}
    for i in range(len(genres)):
        genre = genres[i]
        play = plays[i]
        genre_play[genre] = genre_play.get(genre, 0) + play
        genre_song.setdefault(genre, []).append((i, play))
    
    answer = []
    sorted_genres = sorted(genre_play.items(), key=lambda x: x[1], reverse=True)
    
    for genre, _ in sorted_genres:
        sorted_songs = sorted(genre_song[genre], key=lambda x: (-x[1], x[0]))
        answer.extend([idx for idx, _ in sorted_songs[:2]])
    
    return answer